// serverless-plugin-datadog/dist/env.js
"use strict";
/*
 * Unless explicitly stated otherwise all files in this repository are licensed
 * under the Apache License Version 2.0.
 *
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2019 Datadog, Inc.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const apiKeyEnvVar = "DD_API_KEY";
const apiKeyKMSEnvVar = "DD_KMS_API_KEY";
const siteURLEnvVar = "DD_SITE";
const logLevelEnvVar = "DD_LOG_LEVEL";
const logForwardingEnvVar = "DD_FLUSH_TO_LOG";
exports.defaultConfiguration = {
    addLayers: true,
    flushMetricsToLogs: true,
    logLevel: "info",
    site: "datadoghq.com",
    enableXrayTracing: false,
    enableDDTracing: true,
    enableTags: true,
};
function setEnvConfiguration(config, service) {
    const provider = service.provider;
    if (provider.environment === undefined) {
        provider.environment = {};
    }
    const environment = provider.environment;
    if (config.apiKey !== undefined && environment[apiKeyEnvVar] === undefined) {
        environment[apiKeyEnvVar] = config.apiKey;
    }
    if (config.apiKMSKey !== undefined && environment[apiKeyKMSEnvVar] === undefined) {
        environment[apiKeyKMSEnvVar] = config.apiKMSKey;
    }
    if (environment[siteURLEnvVar] === undefined) {
        environment[siteURLEnvVar] = config.site;
    }
    if (environment[logLevelEnvVar] === undefined) {
        environment[logLevelEnvVar] = config.logLevel;
    }
    if (environment[logForwardingEnvVar] === undefined) {
        environment[logForwardingEnvVar] = config.flushMetricsToLogs;
    }
}
exports.setEnvConfiguration = setEnvConfiguration;
function getConfig(service) {
    let custom = service.custom;
    if (custom === undefined) {
        custom = {};
    }
    let datadog = custom.datadog;
    if (datadog === undefined) {
        datadog = {};
    }
    return Object.assign(Object.assign({}, exports.defaultConfiguration), datadog);
}
exports.getConfig = getConfig;
